File: kraken-engine/src/main/java/kraken/runtime/engine/context/data/DataContextBuilder.java
/*
 *  Copyright 2019 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package kraken.runtime.engine.context.data;

import static kraken.message.SystemMessageBuilder.Message.CONTEXT_BUILD_INVALID_DATA;
import static kraken.utils.MessageUtils.withSpaceBeforeEachLine;

import java.util.stream.Collectors;

import kraken.message.SystemMessageBuilder;
import kraken.runtime.DataContextPathProvider;
import kraken.runtime.engine.context.extraction.ContextChildExtractionInfo;
import kraken.runtime.engine.context.extraction.instance.ContextExtractionResult;
import kraken.runtime.engine.context.info.ContextInstanceInfo;
import kraken.runtime.engine.context.info.ContextInstanceInfoResolver;
import kraken.runtime.engine.context.info.DataErrorDefinition;
import kraken.runtime.repository.RuntimeContextRepository;

/**
 * Creates a {@link DataContext} instance from passed data object.
 *
 * @author rimas
 * @since 1.0
 */
@SuppressWarnings("WeakerAccess")
public class DataContextBuilder {

    private final RuntimeContextRepository contextRepository;
    private final ContextInstanceInfoResolver<?> instanceInfoResolver;
    private final DataContextPathProvider dataContextPathProvider;

    public DataContextBuilder(
            RuntimeContextRepository contextRepository,
            ContextInstanceInfoResolver<?> instanceInfoResolver,
            DataContextPathProvider dataContextPathProvider
    ) {
        this.contextRepository = contextRepository;
        this.instanceInfoResolver = instanceInfoResolver;
        this.dataContextPathProvider = dataContextPathProvider;
    }

    /**
     * Produce {@link DataContext} instance from passed context object instance
     */
    public DataContext buildFromRoot(Object rootContextObject) {
        throwIfContextDataObjectNotValid(rootContextObject);
        return buildContext(rootContextObject, instanceInfoResolver.resolveRootInfo(rootContextObject), null);
    }

    /**
     * Produce ancestor {@link DataContext} instance
     */
    public DataContext buildForAncestorObject(Object contextDataObject,
                                              String ancestorContextName,
                                              DataContext childContext) {
        throwIfContextDataObjectNotValid(contextDataObject);
        ContextInstanceInfo contextInstanceInfo = instanceInfoResolver.resolveAncestorInfo(
            contextDataObject,
            contextRepository.getContextDefinition(ancestorContextName),
            contextRepository.getContextDefinition(childContext.getContextName()),
            childContext.getContextInstanceInfo()
        );
        return buildContext(contextDataObject, contextInstanceInfo, null);
    }

    public DataContext buildFromExtractedObject(ContextExtractionResult instance, ContextChildExtractionInfo info) {
        throwIfContextDataObjectNotValid(instance.getValue());
        ContextInstanceInfo contextInstanceInfo = instanceInfoResolver.resolveExtractedInfo(
            instance.getValue(),
            contextRepository.getContextDefinition(info.getChildContextName()),
            contextRepository.getContextDefinition(info.getParentDataContext().getContextName()),
            info.getParentDataContext().getContextInstanceInfo()
        );
        return buildContext(
            instance.getValue(),
            contextInstanceInfo,
            info.getParentDataContext()
        );
    }

    private DataContext buildContext(Object contextDataObject, ContextInstanceInfo info, DataContext parentDataContext) {
        var dataContext = new DataContext();
        var contextDefinition = contextRepository.getContextDefinition(info.getContextName());

        dataContext.setContextName(info.getContextName());
        dataContext.setContextId(info.getContextInstanceId());
        dataContext.setContextPath(dataContextPathProvider.getPath(info.getContextInstanceId()));
        dataContext.setDataObject(contextDataObject);
        dataContext.setContextDefinition(contextDefinition);
        dataContext.setContextInstanceInfo(info);
        dataContext.setParentDataContext(parentDataContext);

        return dataContext;
    }

    private void throwIfContextDataObjectNotValid(Object contextDataObject) {
        var errors = instanceInfoResolver.validateContextDataObject(contextDataObject);
        if (!errors.isEmpty()) {
            var joinedReasons = errors.stream()
                .map(DataErrorDefinition::getMessage)
                .collect(Collectors.joining(System.lineSeparator()));

            var className = contextDataObject != null ? contextDataObject.getClass().getName() : "null";

            var m = SystemMessageBuilder.create(CONTEXT_BUILD_INVALID_DATA)
                .parameters(className, System.lineSeparator() + withSpaceBeforeEachLine(joinedReasons))
                .build();
            throw new DataContextBuildingException(m);
        }
    }

}

File: javascript/packages/engine/src/engine/contexts/data/DataContextBuilder.ts
/*
 *  Copyright 2018 EIS Ltd and/or one of its affiliates.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

import { ContextInstanceInfoResolver } from '../info/ContextInstanceInfoResolver'
import { requireDefinedValue } from '../../../utils/Utils'
import { ContextModelTree } from '../../../models/ContextModelTree'
import { DataContext } from './DataContext'
import {
    CONTEXT_BUILD_INVALID_DATA,
    CONTEXT_BUILD_UNDEFINED_DATA,
    KrakenRuntimeError,
    SystemMessageBuilder,
    UNKNOWN_CONTEXT_DEFINITION,
} from '../../../error/KrakenRuntimeError'
import { ContextInstanceInfo } from 'kraken-engine-api'

export class DataContextBuilder {
    constructor(
        private readonly modelTree: ContextModelTree.ContextModelTree,
        private readonly resolver: ContextInstanceInfoResolver<unknown>,
    ) {}

    /**
     * Produce {@link DataContext} instance from passed context object instance
     */
    buildFromRoot(rootContextObject: object): DataContext {
        const { requireValid, build, resolver } = this
        return build(requireValid(rootContextObject), resolver.resolveRootInfo(rootContextObject))
    }

    /**
     * Produce {@link DataContext} instance from extracted object
     */
    buildFromExtractedObject(
        contextDataObject: object,
        childContextName: string,
        parentContext: DataContext,
        index?: number,
    ): DataContext {
        const { modelTree, requireValid, build, resolver } = this
        const info = resolver.resolveExtractedInfo(
            requireValid(contextDataObject),
            modelTree.contexts[childContextName],
            modelTree.contexts[parentContext.contextName],
            parentContext.info,
            index,
        )
        return build(contextDataObject, info, parentContext)
    }

    private build = (data: object, info: ContextInstanceInfo, parent?: DataContext): DataContext => {
        requireDefinedValue(info, 'Context instance info is null')
        const contextDefinitionName = requireDefinedValue(info.getContextName(), 'Context definition name is null')
        const contextDefinition = this.modelTree.contexts[contextDefinitionName]

        if (!contextDefinition) {
            const m = new SystemMessageBuilder(UNKNOWN_CONTEXT_DEFINITION).parameters(contextDefinitionName).build()
            throw new KrakenRuntimeError(m)
        }

        return new DataContext(
            requireDefinedValue(info.getContextInstanceId(), 'Context instance id is null'),
            contextDefinitionName,
            data as Record<string, unknown>,
            info,
            contextDefinition,
            parent,
        )
    }

    private requireValid: (data: object) => object = (data: object) => {
        if (data == undefined) {
            const m = new SystemMessageBuilder(CONTEXT_BUILD_UNDEFINED_DATA).build()
            throw new KrakenRuntimeError(m)
        }
        const errors = this.resolver.validateContextDataObject(data)
        if (errors.length) {
            const reasons = errors.map(error => `  ${error.message}`).join('\n')
            const m = new SystemMessageBuilder(CONTEXT_BUILD_INVALID_DATA).parameters('\n' + reasons).build()

            throw new KrakenRuntimeError(m)
        }
        return data
    }
}
